californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -139,39 +159,38 @@ public BlockwiseLayer(final NetworkConfig config) {
 	@Override
 	public void sendRequest(final Exchange exchange, final Request request) {
 
-		BlockOption block2 = request.getOptions().getBlock2();
-		if (block2 != null && block2.getNum() > 0) {
-			// This is the case if the user has explicitly added a block option
-			// for random access.
-			// Note: We do not regard it as random access when the block num is
-			// 0. This is because the user might just want to do early block
-			// size negotiation but actually wants to receive all blocks.
-			LOGGER.fine("request contains block2 option, creating random-access blockwise status");
-			BlockwiseStatus status = new BlockwiseStatus(getSizeForSzx(block2.getSzx()), request.getOptions().getContentFormat());
-			status.setCurrentSzx(block2.getSzx());
-			status.setCurrentNum(block2.getNum());
-			status.setRandomAccess(true);
-			exchange.setResponseBlockStatus(status);
-			lower().sendRequest(exchange, request);
-
-		} else if (requiresBlockwise(request)) {
-			// This must be a large POST or PUT request
-			startBlockwiseUpload(exchange, request);
+		Request requestToSend = request;
 
-		} else {
-			// no blockwise transfer required
-			exchange.setCurrentRequest(request);
-			lower().sendRequest(exchange, request);
+		if (isTransparentBlockwiseHandlingEnabled()) {
+
+			BlockOption block2 = request.getOptions().getBlock2();
+			if (block2 != null && block2.getNum() > 0) {
+				// This is the case if the user has explicitly added a block option
+				// for random access.
+				// Note: We do not regard it as random access when the block number is 0.
+				// This is because the user might just want to do early block
+				// size negotiation but actually want to retrieve the whole body by means of
+				// a transparent blockwise transfer.
+				LOGGER.fine("outbound request contains block2 option, creating random-access blockwise status");
+				addRandomAccessBlock2Status(exchange, request, block2);
+
+			} else if (requiresBlockwise(request)) {
+				// This must be a large POST or PUT request
+				requestToSend = startBlockwiseUpload(exchange, request);
+
+			}
 		}
+
+		exchange.setCurrentRequest(requestToSend);
+		lower().sendRequest(exchange, requestToSend);
 	}
 
-	private void startBlockwiseUpload(final Exchange exchange, final Request request) {
+	private Request startBlockwiseUpload(final Exchange exchange, final Request request) {
 
-		BlockwiseStatus status = findRequestBlockStatus(exchange, request);
+		final KeyUri key = getKey(exchange, request);
+		Block1BlockwiseStatus status = getOutboundBlock1Status(key, exchange, request);
```

Comment:
> Don't you think that the client's intention is to either upload the whole body or nothing at all? If a second PUT/POST (from another thread representing e.g. another client app on top of leshan) now triggers cancelation of the first block1 transfer, this could hardly be called deliberately canceled by client one, could it?

Probably not, but this is because you consider that there are 2 different "entities" (you call it _client app_) using the Californium client API  and sharing the same CoAP server resource without knowing each other.

FMPOV, There is only one user of the Californium client API, which should know what he is doing. In this case we could easily say that this is done deliberately.  If he decide to send a new request before it receive the response to the first one, this means that finally he doesn't need to send the previous request, only the last value make sense to be PUT/POST. 
If it want to support the use case you described, up to him to choose/implement if the first request should be cancelled or the second one rejected or maybe queued.